content/blog/similar-document-search-using-apache-spark-with-tf-idf.md

```diff
@@ -0,0 +1,167 @@
+
+```
+def toJson(v: Vector): String = {
+   v match {
+     case SparseVector(size, indices, values) =>
+       val jValue = ("type" -> 0) ~
+         ("size" -> size) ~
+         ("indices" -> indices.toSeq) ~
+         ("values" -> values.toSeq)
+       compact(render(jValue))
+     case DenseVector(values) =>
+       val jValue = ("type" -> 1) ~ ("values" -> values.toSeq)
+       compact(render(jValue))
+        }
+      }
+}
+val asJsonUDF = udf[String,Vector](toJson)
+```
+
+8. Finally, saves features vector to the MapR Database table.
+
+```
+val dfToSave = normalized.withColumn("rawFeaturesJson", asJsonUDF(col("rawFeatures"))).withColumn("featuresJson", asJsonUDF(col("features"))).drop("rawFeatures").drop("features")
+dfToSave.saveToMapRDB("/googlegroups/trained_model", createTable = false)
+```
+
+## Application 2 - New Messages
+
+The second application is a [Spark Stream Consumer application](https://developer.hpe.com/blog/streaming-machine-learning-pipeline-for-sentiment-analysis-using-apache-/) that **will execute the following steps:**
+
+1. Loads the previously saved `idfModel` and initializes a new `HashingTF` model.
+
+```
+val idfModel = IDFModel.load("path/to/serialized/model")
+val hashingTF = new HashingTF().setInputCol("words").setOutputCol("rawFeatures").setNumFeatures(1000)
+```
+
+2. Loads in memory and caches the data with the features, saved previously.
+
+```
+val all = contextFuntions.loadFromMapRDB(argsConfiguration.trained).toDF
+all.cache()
+```
+
+3. Creates a DataFrame with current message.
+
+```
+val one = Seq((x._id,x.body)).toDF("_id", "contents")
+val newWords = prepareWords(one, "words")
+val newFeature = hashingTF.transform(newWords)
+val newRescale = idfModel.transform(newFeature)
+val normalized = newRescale.withColumn("norm2", UDF.calcNormUDF(col("features2")))
+```
+
+4. Then it finds the crossjoin DataFrame between the one element and all existing messages in the database and calculates the similarity.
+
+```
+val cross = normalized.crossJoin(all).drop(normalized.col("_id"))
+val cosine = cross.withColumn("similarity", UDF.calcCosineUDF(col("features"), col("features2"), col("norm"), col("norm2")))
+```
+
+5. For this, it uses the cosine function implemented as follows and registered as a UDF.
+
+```
+def cosineSimilarity(vectorA: SparseVector, vectorB:SparseVector,normASqrt:Double,normBSqrt:Double) :(Double) = {
+ var dotProduct = 0.0
+ for (i <-  vectorA.indices){ dotProduct += vectorA(i) * vectorB(i) }
+ val div = (normASqrt * normBSqrt)
+ if( div == 0 ) (0)
+ else (dotProduct / div)
+}
+udf[Double,SparseVector,SparseVector,Double,Double](cosineSimilarity)
+```
+
+6. The result can then be ordered by similarity, in descending order, taking the top five elements.
+
+```
+val similarsDF = cosine.sort(desc("similarity")).select("similarity","_id").limit(5)
+```
+
+![](/img/image2.png)
```

Comment:
I am not sure why the image is not loaded correctly.
I propose to arrange a quick call to work together and this issue and try to figure out the reason.